litmus-portal/graphql-server/pkg/k8s/cluster.go

```diff
@@ -71,50 +74,148 @@ func ClusterResource(manifest string, namespace string) (*unstructured.Unstructu
+				if path.Backend.ServiceName == ServerServiceName {
+					f := func(c rune) bool {
+						return c == '/'
+					}
+
+					path_arr := strings.FieldsFunc(path.Path, f)
+					if len(path_arr) > 0 {
+						if path_arr[len(path_arr)-1] == "(.*)" {
+							path_arr[len(path_arr)-1] = "query"
+						} else {
+							path_arr = append(path_arr, "query")
+						}
+					} else {
+						path_arr = append(path_arr, "query")
+					}
+
+					IngressPath = strings.Join(path_arr[:], "/")
+				}
+			}
+		}
 
-	for _, addr := range nodeIP.Status.Addresses {
-		if addr.Type == "ExternalIP" && addr.Address != "" {
-			ExternalIP = addr.Address
-		} else if addr.Type == "InternalIP" && addr.Address != "" {
-			InternalIP = addr.Address
+		if len(getIng.Spec.TLS) > 0 {
+			Scheme = "https"
+		} else {
+			Scheme = "http"
 		}
-	}
 
-	if ExternalIP == "" {
-		return "http://" + InternalIP + ":" + strconv.Itoa(int(Nodeport)), nil
-	} else {
-		return "http://" + ExternalIP + ":" + strconv.Itoa(int(Nodeport)), nil
+		FinalUrl = Scheme + "://" + IPAddress + "/" + IngressPath
+
+	} else if Ingress == "false" || Ingress == "" {
+
+		svc, err := clientset.CoreV1().Services(LitmusPortalNS).Get(ServerServiceName, metaV1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+
+		for _, port := range svc.Spec.Ports {
+			if port.Name == "graphql-server" {
+				NodePort = port.NodePort
+				Port = port.Port
+			}
+		}
+
+		if strings.ToLower(string(svc.Spec.Type)) == "loadbalancer" {
+			log.Print("loadbalancer")
+			IPAddress = svc.Spec.LoadBalancerIP
+			if IPAddress == "" {
+				return "", errors.New("ExternalIP is not present for loadbalancer service type")
+			}
+			FinalUrl = "http://" + IPAddress + ":" + strconv.Itoa(int(Port))
+		} else if strings.ToLower(string(svc.Spec.Type)) == "nodeport" {
+			nodeIP, err := clientset.CoreV1().Nodes().Get(NodeName, metaV1.GetOptions{})
+			if err != nil {
+				return "", err
+			}
+
+			for _, addr := range nodeIP.Status.Addresses {
+				if strings.ToLower(string(addr.Type)) == "externalip" && addr.Address != "" {
+					IPAddress = addr.Address
+				} else if strings.ToLower(string(addr.Type)) == "internalip" && addr.Address != "" {
+					InternalIP = addr.Address
+				}
+			}
+
+			if IPAddress == "" {
+				FinalUrl = "http://" + InternalIP + ":" + strconv.Itoa(int(NodePort))
+			} else if InternalIP == "" {
+				FinalUrl = "http://" + IPAddress + ":" + strconv.Itoa(int(NodePort))
+			} else {
+				return "", errors.New("Both ExternalIP and InternalIP aren't present for NodePort service type")
+			}
+		} else if strings.ToLower(string(svc.Spec.Type)) == "clusterip" {
+			log.Print("External agents can't be connected to the server if the service type is set to ClusterIP\n")
+			if svc.Spec.ClusterIP == "" {
+				return "", errors.New("ClusterIP is not present")
+			}
+			FinalUrl = "http://" + svc.Spec.ClusterIP + ":" + strconv.Itoa(int(Port))
+		} else {
+			return "", errors.New("No service type found")
```

Comment:
we can convert to switch case here, it might be more suitable for this type of equality checks

Reply:
yes `strings.ToLower(string(svc.Spec.Type))` operation is repeating @rajdas98

Reply:
Done